modules/login/php/endpoints/authentication.class.inc

```diff
@@ -0,0 +1,453 @@
+        }
+        $requestAccountData['site'] = \Utility::getSiteList();
+        $values['requestAccount']   = $requestAccountData;
+
+        return $values;
+    }
+
+    /**
+     * Reset password.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function resetPassword(array $values) : array
+    {
+        // Always claim success to avoid providing user info to attackers.
+        $values['success'] = 'Password reset. '
+            . 'You should receive an email within a few minutes.';
+
+        // Create the user object.
+        $username = $values['username'];
+        $user     =& \User::singleton($username);
+
+        // Check if valid user.
+        if ($user->userID) {
+            $email = $user->getData('Email');
+            if (empty($email)) {
+                $values['error'] = 'Not a valid user';
+                return $values;
+            }
+
+            try {
+                // Check if valid email.
+                if ($user->isEmailValid()) {
+                    $config = \NDB_Config::singleton();
+                    // Generate a new password.
+                    $password = \Utility::randomString();
+                    // Reset the password in the database and
+                    // expire password so user must change it upon login.
+                    $user->updatePassword(
+                        new \Password($password),
+                        new \DateTime('1999-01-01')
+                    );
+                    // Send the user an email.
+                    $msg_data          = array();
+                    $msg_data['study'] = $config->getSetting('title');
+                    $msg_data['url']   = $config->getSetting('url');
+                    $msg_data['realname'] = $user->getData('Real_name');
+                    $msg_data['password'] = $password;
+                    \Email::send($email, 'lost_password.tpl', $msg_data);
+                } else {
+                    $values['error'] = 'Sending reset email failed.';
+                    error_log(
+                        $_SERVER['REMOTE_ADDR']
+                        . '(' . $_SERVER['HTTP_X_FORWARDED_FOR'] . ')'
+                        . ' attempted to reset password for ' . $username
+                    );
+                }
+            } catch (\LorisException $e) {
+                error_log($e->getMessage());
+                $values['error'] = 'Sending reset email failed.';
+            }
+        }
+        return $values;
+    }
+
+    /**
+     * Process request account form.
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     */
+    function requestAccount(array $values) : array
+    {
+        $config           =& \NDB_Config::singleton();
+        $reCAPTCHAPrivate = $config->getSetting('reCAPTCHAPrivate');
+
+        // Verify reCAPTCHA
+        if (isset($values['g-recaptcha-response']) && isset($reCAPTCHAPrivate)) {
```

Comment:
An error should probably thrown if these values aren't set in the request. Right now you could avoid this CAPTCHA check by submitting a request manually because the back-end doesn't do any validation.